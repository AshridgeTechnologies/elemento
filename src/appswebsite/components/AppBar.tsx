import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}