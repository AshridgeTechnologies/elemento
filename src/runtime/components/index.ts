export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
