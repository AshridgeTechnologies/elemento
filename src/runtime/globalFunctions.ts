import {fromPairs, isNil, last, sort, sortBy, splitEvery} from 'ramda'
import {
    add,
    differenceInCalendarDays,
    differenceInDays,
    differenceInHours,
    differenceInMinutes,
    differenceInMonths,
    differenceInSeconds,
    differenceInYears,
    format,
    parseISO
} from 'date-fns'
import * as csv from 'csv-parse/sync'
import {Value, valueOf, valuesOf} from './runtimeFunctions'
import {isNumeric, noSpaces} from '../util/helpers'
import {CastingContext} from 'csv-parse/lib'

type TimeUnit = 'seconds' | 'minutes' | 'hours' | 'days' | 'months' | 'years'
const unitTypes = ['seconds' , 'minutes' , 'hours' , 'days' , 'months' , 'years']

export const globalFunctions = {
    valueOf,

    Sum(...args: Value<number>[]) {
        return args.reduce( (acc, val) => (acc as number) + (val as number), 0)
    },

    Log(...args: any[]) {
        console.log(...args)
    },

    If(condition: any, trueValue: any, falseValue?: any) {
        return valueOf(condition) ? trueValue : falseValue
    },

    Left(s: Value<string>, length: Value<number>) {
        return valueOf(s).substring(0, valueOf(length))
    },

    Mid(s: Value<string>, start: Value<number>, length?: Value<number>) {
        const sVal = valueOf(s), startVal = valueOf(start), lengthVal = valueOf(length)
        if (startVal < 1) throw new Error(`Function Mid parameter 2 (start) is ${startVal}. It should be greater than or equal to 1.`)
        if (lengthVal !== undefined && lengthVal < 0) throw new Error(`Function Mid parameter 3 (length) is ${lengthVal}. It should be greater than or equal to 0.`)
        return sVal.substring(startVal - 1, startVal - 1 + (lengthVal ?? sVal.length))
    },

    Right(s: Value<string>, length: Value<number>) {
        const sVal = valueOf(s), lengthVal = valueOf(length)
        return sVal.substring(sVal.length - lengthVal)
    },

    And(...args: Value<any>[]) {
        return args.reduce( (prev, curr) => prev && !!valueOf(curr), true )
    },

    Or(...args: Value<any>[]) {
        return args.reduce( (prev, curr) => prev || !!valueOf(curr), false )
    },

    Not(arg: Value<any>) {
        return !valueOf(arg)
    },

    Substitute(s: Value<string>, toReplace: Value<string>, replaceWith: Value<string>) {
        const sVal = valueOf(s), toReplaceVal = valueOf(toReplace), replaceWithVal = valueOf(replaceWith)
        if (toReplaceVal === '') return sVal
        return sVal.replace(new RegExp(toReplaceVal, 'g'), replaceWithVal)
    },

    Max(...args: Value<number>[]) {
        if (args.length === 0) throw new Error('Wrong number of arguments to Max. Expected at least 1 argument.')
        return Math.max(...args as number[])
    },

    Min(...args: Value<number>[]) {
        if (args.length === 0) throw new Error('Wrong number of arguments to Min. Expected at least 1 argument.')
        return Math.min(...args as number[])
    },

    Record(...args: Value<any>[]) {
        if (args.length % 2 !== 0) throw new Error('Odd number of arguments - must have pairs of name, value')
        const argVals = args.map( valueOf )
        const pairs = splitEvery(2, argVals) as [string, any][]
        return fromPairs(pairs)
    },

    List(...args: Value<any>[]) {
        return args.map( valueOf )
    },

    Select(list: Value<any[]>, condition: (item: any) => boolean) {
        const listVal = valueOf(list) ?? []
        if (condition === undefined) throw new Error('Wrong number of arguments to Select. Expected list, expression.')
        return listVal.filter(condition)
    },

    ForEach(list: Value<any[]>, transform: (item: any) => any) {
        const listVal = valueOf(list) ?? []
        if (transform === undefined) throw new Error('Wrong number of arguments to ForEach. Expected list, expression.')
        return listVal.map(transform)
    },

    First(list: Value<any[]>, condition: (item: any) => boolean = () => true) {
        const listVal = valueOf(list) ?? []
        if (list === undefined) throw new Error('Wrong number of arguments to First. Expected list, optional expression.')
        return listVal.filter(condition)[0] ?? null
    },

    Last(list: Value<any[]>, condition: (item: any) => boolean = () => true) {
        const listVal = valueOf(list) ?? []
        if (list === undefined) throw new Error('Wrong number of arguments to Last. Expected list, optional expression.')
        return last(listVal.filter(condition)) ?? null
    },

    Sort(list: Value<any[]>, sortKeyFn: (item: any) => any | any[]): any[] {
        const compareItems = (a: any, b: any): -1 | 0 | 1 => {
            const aa = sortKeyFn(a), bb = sortKeyFn(b)
            return compareValues(aa, bb)
        }

        const compareValues = (a: any, b: any): -1 | 0 | 1 => {
            if (isNil(a) && isNil(b)) {
                return 0
            } else if (isNil(a)) {
                return -1
            } else if (isNil(b)) {
                return 1
            }
            if (Array.isArray(a) && Array.isArray(b)) {
                return compareArrays(a, b)
            }
            return a < b ? -1 : a > b ? 1 : 0
        }

        const compareArrays = (a: any[], b: any[]) => {
            for (let i = 0; i < a.length; i++) {
                const compareVal = compareValues(a[i], b[i])
                if (compareVal !== 0) {
                    return compareVal
                }
            }
            return 0
        }

        const listVal = valueOf(list)
        if (isNil(listVal)) {
            return listVal
        }
        return sort(compareItems, listVal)
    },

    Timestamp() {
        return Date.now()
    },

    Now() {
        return new Date()
    },

    Today(date = new Date()) {
        return new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate() ))
    },

    TimeBetween(date1Val: Value<Date>, date2Val: Value<Date>, unitVal: Value<TimeUnit>) {
        const [date1, date2, unit] = valuesOf(date1Val, date2Val, unitVal)
        if (!date1 || !date2) return undefined
        const unitTypes = ['seconds' , 'minutes' , 'hours' , 'days' , 'months' , 'years']
        switch(unit) {
            case 'seconds': return differenceInSeconds(date2, date1)
            case 'minutes': return differenceInMinutes(date2, date1)
            case 'hours': return differenceInHours(date2, date1)
            case 'days': return differenceInDays(date2, date1)
            case 'months': return differenceInMonths(date2, date1)
            case 'years': return differenceInYears(date2, date1)
            default: throw new Error(`Unknown unit ${unit}.  Should be one of ${unitTypes.join(', ')}`)
        }

    },

    DaysBetween(date1Val: Value<Date>, date2Val: Value<Date>) {
        const [date1, date2] = valuesOf(date1Val, date2Val)
        if (!date1 || !date2) return undefined
        return differenceInCalendarDays(date2, date1)
    },

    DateAdd(dateVal: Value<Date>, changeVal: Value<number>, unitVal: Value<TimeUnit>) {
        const [date, change, unit] = valuesOf(dateVal, changeVal, unitVal)
        if (!unitTypes.includes(unit)) {
            throw new Error(`Unknown unit ${unit}.  Should be one of ${unitTypes.join(', ')}`)
        }
        if (!date || change === undefined || change === null) return undefined

        return add(date, {[unit]: change})
    },

    DateFormat(date: Value<Date>, pattern: string) {
        if (!date) return ''
        return format(valueOf(date), pattern)
    },

    CsvToRecords(csvText: string, columnNames?: string[]) {
        const cast = (field: string, context: CastingContext) => {
            if (context.quoting) return field
            if (field === '') return undefined
            if (isNumeric(field)) return Number(field)
            if (field.toLowerCase() === 'true') return true
            if (field.toLowerCase() === 'false') return false
            const date = parseISO(field)
            if (!Number.isNaN(date.getTime())) {
                return date
            }
            return field
        }
        const convertColumnNames = (names: string[]) => names.map( noSpaces )
        const removeUndefinedFields = (obj: object) => {
            const definedEntries = Object.entries(obj).filter( ([, value]) => value !== undefined)
            return Object.fromEntries(definedEntries)
        }
        const lines = csvText.split(/[\r\n]+/)
        const tabDelimiters = lines.every( line => line.includes('\t'))
        return csv.parse(csvText, {
            columns: columnNames ? convertColumnNames(columnNames): convertColumnNames,
            delimiter: tabDelimiters ? '\t' : ',',
            cast,
            skip_empty_lines: true,
            skip_records_with_empty_values: true,
            relax_quotes: true,
            relax_column_count: true,
            trim: true
        }).map(removeUndefinedFields)
    }
}

export const functionArgIndexes = {
    Select: [1],
    ForEach: [1],
    First: [1],
    Last: [1],
    Sort: [1],
}