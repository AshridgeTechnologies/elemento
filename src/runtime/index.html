<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>