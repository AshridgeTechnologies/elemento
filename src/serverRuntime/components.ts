export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts