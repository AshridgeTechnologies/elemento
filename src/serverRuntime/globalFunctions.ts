export {globalFunctions} from '../runtime/globalFunctions'
