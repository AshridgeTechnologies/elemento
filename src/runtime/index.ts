import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
