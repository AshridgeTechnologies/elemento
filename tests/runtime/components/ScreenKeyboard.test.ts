/**
 * @vitest-environment jsdom
 */
import {expect, test, vi} from 'vitest'
import {createElement} from 'react'
import {ScreenKeyboard} from '../../../src/runtime/components/index'
import {valueObj, wrappedTestElementNew} from '../../testutil/testHelpers'
import {globalFunctions} from '../../../src/runtime/globalFunctions'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import {render} from '@testing-library/react'

const {Log} = globalFunctions

const doIt = () => {}

const [screenKeyboard, appStoreHook] = wrappedTestElementNew(ScreenKeyboard)

const stateAt = (path: string) => appStoreHook.stateAt(path)

let container: any

test('ScreenKeyboard element produces output with properties supplied', async () => {
    await actWait( () => ({container} = render(screenKeyboard('app.page1.keyboard', {}))))
    expect(container.innerHTML).toMatchSnapshot()
})

test('ScreenKeyboard element produces output with styles supplied', async () => {
    await actWait(() => ({container} = render(screenKeyboard('app.page1.keyboard', {styles: {width: 300}}))))
    expect(container.innerHTML).toMatchSnapshot()
})

test('ScreenKeyboard element produces output with show false', async () => {
    await actWait(() => ({container} = render(screenKeyboard('app.page1.keyboard', {show: false}))))
    expect(container.innerHTML).toMatchSnapshot()
})

test('ScreenKeyboard element produces output with properties supplied as state values', async () => {
    await actWait(() => ({container} = render(screenKeyboard('app.page1.keyboard', {styles: {width: valueObj(300)}}))))
    expect(container.innerHTML).toMatchSnapshot()
})

test('ScreenKeyboard does key action when clicked and translates keys', async () => {
    let {element, user} = testContainer(screenKeyboard('app.page1.keyboard', {keyAction: ($key: string) => Log("Pressed", $key)}))
    const keyEl = (key: string) => element(`div[data-skbtn="${key}"]`)
    const log = vi.spyOn(console, "log").mockImplementation(() => {})
    try {
        await user.click(keyEl('g'))
        expect(log).toHaveBeenLastCalledWith("Pressed", 'g')
        await user.click(keyEl('{ent}'))
        expect(log).toHaveBeenLastCalledWith("Pressed", 'Enter')
        await user.click(keyEl('{backspace}'))
        expect(log).toHaveBeenLastCalledWith("Pressed", 'Backspace')
    } finally {
        log.mockReset();
    }
} )

test('ScreenKeyboard stores value from clicked keys and can Reset', async () => {
    let {element, user} = testContainer(screenKeyboard('app.page1.keyboard', {}))
    const keyEl = (key: string) => element(`div[data-skbtn="${key}"]`)
    await user.click(keyEl('a'))
    await user.click(keyEl('b'))
    await user.click(keyEl('c'))
    expect(stateAt('app.page1.keyboard').value).toBe('abc')
    await user.click(keyEl('{backspace}'))
    expect(stateAt('app.page1.keyboard').value).toBe('ab')

    await actWait( ()=> stateAt('app.page1.keyboard').Reset() )
    expect(stateAt('app.page1.keyboard').value).toBe('')
} )

test('ScreenKeyboard uses real keyboard', async () => {
    let {element, user} = testContainer(
        createElement('div', {className: 'ElPage', tabIndex: 0}, screenKeyboard('app.page1.keyboard', {useRealKeyboard: true}))
    )
    const pageEl = element(`div.ElPage`)
    pageEl.focus()
    const log = vi.spyOn(console, "log").mockImplementation(() => {})
    try {
        await user.keyboard('a')
        await user.keyboard('b')
        expect(stateAt('app.page1.keyboard').value).toBe('ab')
        await user.keyboard('{Enter}')
        expect(stateAt('app.page1.keyboard').value).toBe('ab')
        await user.keyboard('{Backspace}')
        expect(stateAt('app.page1.keyboard').value).toBe('a')
    } finally {
        log.mockReset();
    }
} )

