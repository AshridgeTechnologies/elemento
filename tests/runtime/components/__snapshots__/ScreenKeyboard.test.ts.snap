// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenKeyboard element produces output with properties supplied 1`] = `"<div class="MuiBox-root css-0" id="app.page1.keyboard" tabindex="-1"><div class="react-simple-keyboard simple-keyboard hg-theme-default hg-layout-lettersOnly" data-skinstance="reactSimpleKeyboard"><div class="hg-rows"><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="q" data-skbtnuid="lettersOnly-r0b0"><span>q</span></div><div class="hg-button hg-standardBtn" data-skbtn="w" data-skbtnuid="lettersOnly-r0b1"><span>w</span></div><div class="hg-button hg-standardBtn" data-skbtn="e" data-skbtnuid="lettersOnly-r0b2"><span>e</span></div><div class="hg-button hg-standardBtn" data-skbtn="r" data-skbtnuid="lettersOnly-r0b3"><span>r</span></div><div class="hg-button hg-standardBtn" data-skbtn="t" data-skbtnuid="lettersOnly-r0b4"><span>t</span></div><div class="hg-button hg-standardBtn" data-skbtn="y" data-skbtnuid="lettersOnly-r0b5"><span>y</span></div><div class="hg-button hg-standardBtn" data-skbtn="u" data-skbtnuid="lettersOnly-r0b6"><span>u</span></div><div class="hg-button hg-standardBtn" data-skbtn="i" data-skbtnuid="lettersOnly-r0b7"><span>i</span></div><div class="hg-button hg-standardBtn" data-skbtn="o" data-skbtnuid="lettersOnly-r0b8"><span>o</span></div><div class="hg-button hg-standardBtn" data-skbtn="p" data-skbtnuid="lettersOnly-r0b9"><span>p</span></div></div><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="a" data-skbtnuid="lettersOnly-r1b0"><span>a</span></div><div class="hg-button hg-standardBtn" data-skbtn="s" data-skbtnuid="lettersOnly-r1b1"><span>s</span></div><div class="hg-button hg-standardBtn" data-skbtn="d" data-skbtnuid="lettersOnly-r1b2"><span>d</span></div><div class="hg-button hg-standardBtn" data-skbtn="f" data-skbtnuid="lettersOnly-r1b3"><span>f</span></div><div class="hg-button hg-standardBtn" data-skbtn="g" data-skbtnuid="lettersOnly-r1b4"><span>g</span></div><div class="hg-button hg-standardBtn" data-skbtn="h" data-skbtnuid="lettersOnly-r1b5"><span>h</span></div><div class="hg-button hg-standardBtn" data-skbtn="j" data-skbtnuid="lettersOnly-r1b6"><span>j</span></div><div class="hg-button hg-standardBtn" data-skbtn="k" data-skbtnuid="lettersOnly-r1b7"><span>k</span></div><div class="hg-button hg-standardBtn" data-skbtn="l" data-skbtnuid="lettersOnly-r1b8"><span>l</span></div></div><div class="hg-row"><div class="hg-button hg-functionBtn hg-button-ent" data-skbtn="{ent}" data-skbtnuid="lettersOnly-r2b0"><span>Enter</span></div><div class="hg-button hg-standardBtn" data-skbtn="z" data-skbtnuid="lettersOnly-r2b1"><span>z</span></div><div class="hg-button hg-standardBtn" data-skbtn="x" data-skbtnuid="lettersOnly-r2b2"><span>x</span></div><div class="hg-button hg-standardBtn" data-skbtn="c" data-skbtnuid="lettersOnly-r2b3"><span>c</span></div><div class="hg-button hg-standardBtn" data-skbtn="v" data-skbtnuid="lettersOnly-r2b4"><span>v</span></div><div class="hg-button hg-standardBtn" data-skbtn="b" data-skbtnuid="lettersOnly-r2b5"><span>b</span></div><div class="hg-button hg-standardBtn" data-skbtn="n" data-skbtnuid="lettersOnly-r2b6"><span>n</span></div><div class="hg-button hg-standardBtn" data-skbtn="m" data-skbtnuid="lettersOnly-r2b7"><span>m</span></div><div class="hg-button hg-functionBtn hg-button-backspace" data-skbtn="{backspace}" data-skbtnuid="lettersOnly-r2b8"><span>⌫</span></div></div></div></div></div>"`;

exports[`ScreenKeyboard element produces output with properties supplied as state values 1`] = `"<div class="MuiBox-root css-1v5z18m" id="app.page1.keyboard" tabindex="-1"><div class="react-simple-keyboard simple-keyboard hg-theme-default hg-layout-lettersOnly" data-skinstance="reactSimpleKeyboard"><div class="hg-rows"><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="q" data-skbtnuid="lettersOnly-r0b0"><span>q</span></div><div class="hg-button hg-standardBtn" data-skbtn="w" data-skbtnuid="lettersOnly-r0b1"><span>w</span></div><div class="hg-button hg-standardBtn" data-skbtn="e" data-skbtnuid="lettersOnly-r0b2"><span>e</span></div><div class="hg-button hg-standardBtn" data-skbtn="r" data-skbtnuid="lettersOnly-r0b3"><span>r</span></div><div class="hg-button hg-standardBtn" data-skbtn="t" data-skbtnuid="lettersOnly-r0b4"><span>t</span></div><div class="hg-button hg-standardBtn" data-skbtn="y" data-skbtnuid="lettersOnly-r0b5"><span>y</span></div><div class="hg-button hg-standardBtn" data-skbtn="u" data-skbtnuid="lettersOnly-r0b6"><span>u</span></div><div class="hg-button hg-standardBtn" data-skbtn="i" data-skbtnuid="lettersOnly-r0b7"><span>i</span></div><div class="hg-button hg-standardBtn" data-skbtn="o" data-skbtnuid="lettersOnly-r0b8"><span>o</span></div><div class="hg-button hg-standardBtn" data-skbtn="p" data-skbtnuid="lettersOnly-r0b9"><span>p</span></div></div><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="a" data-skbtnuid="lettersOnly-r1b0"><span>a</span></div><div class="hg-button hg-standardBtn" data-skbtn="s" data-skbtnuid="lettersOnly-r1b1"><span>s</span></div><div class="hg-button hg-standardBtn" data-skbtn="d" data-skbtnuid="lettersOnly-r1b2"><span>d</span></div><div class="hg-button hg-standardBtn" data-skbtn="f" data-skbtnuid="lettersOnly-r1b3"><span>f</span></div><div class="hg-button hg-standardBtn" data-skbtn="g" data-skbtnuid="lettersOnly-r1b4"><span>g</span></div><div class="hg-button hg-standardBtn" data-skbtn="h" data-skbtnuid="lettersOnly-r1b5"><span>h</span></div><div class="hg-button hg-standardBtn" data-skbtn="j" data-skbtnuid="lettersOnly-r1b6"><span>j</span></div><div class="hg-button hg-standardBtn" data-skbtn="k" data-skbtnuid="lettersOnly-r1b7"><span>k</span></div><div class="hg-button hg-standardBtn" data-skbtn="l" data-skbtnuid="lettersOnly-r1b8"><span>l</span></div></div><div class="hg-row"><div class="hg-button hg-functionBtn hg-button-ent" data-skbtn="{ent}" data-skbtnuid="lettersOnly-r2b0"><span>Enter</span></div><div class="hg-button hg-standardBtn" data-skbtn="z" data-skbtnuid="lettersOnly-r2b1"><span>z</span></div><div class="hg-button hg-standardBtn" data-skbtn="x" data-skbtnuid="lettersOnly-r2b2"><span>x</span></div><div class="hg-button hg-standardBtn" data-skbtn="c" data-skbtnuid="lettersOnly-r2b3"><span>c</span></div><div class="hg-button hg-standardBtn" data-skbtn="v" data-skbtnuid="lettersOnly-r2b4"><span>v</span></div><div class="hg-button hg-standardBtn" data-skbtn="b" data-skbtnuid="lettersOnly-r2b5"><span>b</span></div><div class="hg-button hg-standardBtn" data-skbtn="n" data-skbtnuid="lettersOnly-r2b6"><span>n</span></div><div class="hg-button hg-standardBtn" data-skbtn="m" data-skbtnuid="lettersOnly-r2b7"><span>m</span></div><div class="hg-button hg-functionBtn hg-button-backspace" data-skbtn="{backspace}" data-skbtnuid="lettersOnly-r2b8"><span>⌫</span></div></div></div></div></div>"`;

exports[`ScreenKeyboard element produces output with show false 1`] = `"<div class="MuiBox-root css-1hyfx7x" id="app.page1.keyboard" tabindex="-1"><div class="react-simple-keyboard simple-keyboard hg-theme-default hg-layout-lettersOnly" data-skinstance="reactSimpleKeyboard"><div class="hg-rows"><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="q" data-skbtnuid="lettersOnly-r0b0"><span>q</span></div><div class="hg-button hg-standardBtn" data-skbtn="w" data-skbtnuid="lettersOnly-r0b1"><span>w</span></div><div class="hg-button hg-standardBtn" data-skbtn="e" data-skbtnuid="lettersOnly-r0b2"><span>e</span></div><div class="hg-button hg-standardBtn" data-skbtn="r" data-skbtnuid="lettersOnly-r0b3"><span>r</span></div><div class="hg-button hg-standardBtn" data-skbtn="t" data-skbtnuid="lettersOnly-r0b4"><span>t</span></div><div class="hg-button hg-standardBtn" data-skbtn="y" data-skbtnuid="lettersOnly-r0b5"><span>y</span></div><div class="hg-button hg-standardBtn" data-skbtn="u" data-skbtnuid="lettersOnly-r0b6"><span>u</span></div><div class="hg-button hg-standardBtn" data-skbtn="i" data-skbtnuid="lettersOnly-r0b7"><span>i</span></div><div class="hg-button hg-standardBtn" data-skbtn="o" data-skbtnuid="lettersOnly-r0b8"><span>o</span></div><div class="hg-button hg-standardBtn" data-skbtn="p" data-skbtnuid="lettersOnly-r0b9"><span>p</span></div></div><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="a" data-skbtnuid="lettersOnly-r1b0"><span>a</span></div><div class="hg-button hg-standardBtn" data-skbtn="s" data-skbtnuid="lettersOnly-r1b1"><span>s</span></div><div class="hg-button hg-standardBtn" data-skbtn="d" data-skbtnuid="lettersOnly-r1b2"><span>d</span></div><div class="hg-button hg-standardBtn" data-skbtn="f" data-skbtnuid="lettersOnly-r1b3"><span>f</span></div><div class="hg-button hg-standardBtn" data-skbtn="g" data-skbtnuid="lettersOnly-r1b4"><span>g</span></div><div class="hg-button hg-standardBtn" data-skbtn="h" data-skbtnuid="lettersOnly-r1b5"><span>h</span></div><div class="hg-button hg-standardBtn" data-skbtn="j" data-skbtnuid="lettersOnly-r1b6"><span>j</span></div><div class="hg-button hg-standardBtn" data-skbtn="k" data-skbtnuid="lettersOnly-r1b7"><span>k</span></div><div class="hg-button hg-standardBtn" data-skbtn="l" data-skbtnuid="lettersOnly-r1b8"><span>l</span></div></div><div class="hg-row"><div class="hg-button hg-functionBtn hg-button-ent" data-skbtn="{ent}" data-skbtnuid="lettersOnly-r2b0"><span>Enter</span></div><div class="hg-button hg-standardBtn" data-skbtn="z" data-skbtnuid="lettersOnly-r2b1"><span>z</span></div><div class="hg-button hg-standardBtn" data-skbtn="x" data-skbtnuid="lettersOnly-r2b2"><span>x</span></div><div class="hg-button hg-standardBtn" data-skbtn="c" data-skbtnuid="lettersOnly-r2b3"><span>c</span></div><div class="hg-button hg-standardBtn" data-skbtn="v" data-skbtnuid="lettersOnly-r2b4"><span>v</span></div><div class="hg-button hg-standardBtn" data-skbtn="b" data-skbtnuid="lettersOnly-r2b5"><span>b</span></div><div class="hg-button hg-standardBtn" data-skbtn="n" data-skbtnuid="lettersOnly-r2b6"><span>n</span></div><div class="hg-button hg-standardBtn" data-skbtn="m" data-skbtnuid="lettersOnly-r2b7"><span>m</span></div><div class="hg-button hg-functionBtn hg-button-backspace" data-skbtn="{backspace}" data-skbtnuid="lettersOnly-r2b8"><span>⌫</span></div></div></div></div></div>"`;

exports[`ScreenKeyboard element produces output with styles supplied 1`] = `"<div class="MuiBox-root css-1v5z18m" id="app.page1.keyboard" tabindex="-1"><div class="react-simple-keyboard simple-keyboard hg-theme-default hg-layout-lettersOnly" data-skinstance="reactSimpleKeyboard"><div class="hg-rows"><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="q" data-skbtnuid="lettersOnly-r0b0"><span>q</span></div><div class="hg-button hg-standardBtn" data-skbtn="w" data-skbtnuid="lettersOnly-r0b1"><span>w</span></div><div class="hg-button hg-standardBtn" data-skbtn="e" data-skbtnuid="lettersOnly-r0b2"><span>e</span></div><div class="hg-button hg-standardBtn" data-skbtn="r" data-skbtnuid="lettersOnly-r0b3"><span>r</span></div><div class="hg-button hg-standardBtn" data-skbtn="t" data-skbtnuid="lettersOnly-r0b4"><span>t</span></div><div class="hg-button hg-standardBtn" data-skbtn="y" data-skbtnuid="lettersOnly-r0b5"><span>y</span></div><div class="hg-button hg-standardBtn" data-skbtn="u" data-skbtnuid="lettersOnly-r0b6"><span>u</span></div><div class="hg-button hg-standardBtn" data-skbtn="i" data-skbtnuid="lettersOnly-r0b7"><span>i</span></div><div class="hg-button hg-standardBtn" data-skbtn="o" data-skbtnuid="lettersOnly-r0b8"><span>o</span></div><div class="hg-button hg-standardBtn" data-skbtn="p" data-skbtnuid="lettersOnly-r0b9"><span>p</span></div></div><div class="hg-row"><div class="hg-button hg-standardBtn" data-skbtn="a" data-skbtnuid="lettersOnly-r1b0"><span>a</span></div><div class="hg-button hg-standardBtn" data-skbtn="s" data-skbtnuid="lettersOnly-r1b1"><span>s</span></div><div class="hg-button hg-standardBtn" data-skbtn="d" data-skbtnuid="lettersOnly-r1b2"><span>d</span></div><div class="hg-button hg-standardBtn" data-skbtn="f" data-skbtnuid="lettersOnly-r1b3"><span>f</span></div><div class="hg-button hg-standardBtn" data-skbtn="g" data-skbtnuid="lettersOnly-r1b4"><span>g</span></div><div class="hg-button hg-standardBtn" data-skbtn="h" data-skbtnuid="lettersOnly-r1b5"><span>h</span></div><div class="hg-button hg-standardBtn" data-skbtn="j" data-skbtnuid="lettersOnly-r1b6"><span>j</span></div><div class="hg-button hg-standardBtn" data-skbtn="k" data-skbtnuid="lettersOnly-r1b7"><span>k</span></div><div class="hg-button hg-standardBtn" data-skbtn="l" data-skbtnuid="lettersOnly-r1b8"><span>l</span></div></div><div class="hg-row"><div class="hg-button hg-functionBtn hg-button-ent" data-skbtn="{ent}" data-skbtnuid="lettersOnly-r2b0"><span>Enter</span></div><div class="hg-button hg-standardBtn" data-skbtn="z" data-skbtnuid="lettersOnly-r2b1"><span>z</span></div><div class="hg-button hg-standardBtn" data-skbtn="x" data-skbtnuid="lettersOnly-r2b2"><span>x</span></div><div class="hg-button hg-standardBtn" data-skbtn="c" data-skbtnuid="lettersOnly-r2b3"><span>c</span></div><div class="hg-button hg-standardBtn" data-skbtn="v" data-skbtnuid="lettersOnly-r2b4"><span>v</span></div><div class="hg-button hg-standardBtn" data-skbtn="b" data-skbtnuid="lettersOnly-r2b5"><span>b</span></div><div class="hg-button hg-standardBtn" data-skbtn="n" data-skbtnuid="lettersOnly-r2b6"><span>n</span></div><div class="hg-button hg-standardBtn" data-skbtn="m" data-skbtnuid="lettersOnly-r2b7"><span>m</span></div><div class="hg-button hg-functionBtn hg-button-backspace" data-skbtn="{backspace}" data-skbtnuid="lettersOnly-r2b8"><span>⌫</span></div></div></div></div></div>"`;
